<a href="https://colab.research.google.com/github/Zulabat/My-Python-Library/blob/main/ClimateChange_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [35]:
#Developed by Zula Battulga
#MSIS510 Case 1
#Purple Cohort

import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
from sklearn.metrics import r2_score

In [49]:
#a.i

df_train = 'https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTrain.csv'
df_test = 'https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTest.csv'

train_data = pd.read_csv(df_train)
test_data = pd.read_csv(df_test)

X_train = train_data.drop(columns=['Year', 'Month', 'Temp'])
y_train = train_data['Temp']
X_test = test_data.drop(columns=['Year', 'Month', 'Temp'])
y_test = test_data['Temp']

formula = 'Temp ~ MEI + CO2 + CH4 + N2O + CFC11 + CFC12 + TSI + Aerosols'

linear_regression_model = smf.ols(formula=formula, data=train_data)
linear_regression_model_results = linear_regression_model.fit()

print(linear_regression_model_results.summary())

# Make predictions on the test set
y_pred = linear_regression_model_results.predict(X_test)

# Evaluate model performance using R^2
r2 = r2_score(y_test, y_pred)
print(f'R-squared on test set: {r2}')

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     103.6
Date:                Sun, 13 Oct 2024   Prob (F-statistic):           1.94e-78
Time:                        22:35:51   Log-Likelihood:                 280.10
No. Observations:                 284   AIC:                            -542.2
Df Residuals:                     275   BIC:                            -509.4
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -124.5943     19.887     -6.265      0.0

In [ ]:
#a.ii.
#R-Square Value =  0.751
#By looking at the P value thats are less than 0.05 - MEI, CO2, CFC11, CFC12, TSI, Aerosols are significant independent variable

In [37]:
#a.iii.
#The negative coefficients for N2O and CFC-11 may result from multicollinearity,
#where gases like CO2, CH4, and aerosols are highly correlated and overlap in their effects.

In [45]:
#a.iv.
independent_vars = train_data.drop(columns=['Temp'])
correlation_matrix = independent_vars.corr()

print(correlation_matrix)
#Hightly correlated independent variable for N2O & CFC11.
#N2O - Year, CO2, CH4, CFC12
#CFC11 - CH4, CFC12

              Year     Month       MEI       CO2       CH4       N2O  \
Year      1.000000 -0.027942 -0.036988  0.982749  0.915659  0.993845   
Month    -0.027942  1.000000  0.000885 -0.106732  0.018569  0.013632   
MEI      -0.036988  0.000885  1.000000 -0.041147 -0.033419 -0.050820   
CO2       0.982749 -0.106732 -0.041147  1.000000  0.877280  0.976720   
CH4       0.915659  0.018569 -0.033419  0.877280  1.000000  0.899839   
N2O       0.993845  0.013632 -0.050820  0.976720  0.899839  1.000000   
CFC11     0.569106 -0.013111  0.069000  0.514060  0.779904  0.522477   
CFC12     0.897012  0.000675  0.008286  0.852690  0.963616  0.867931   
TSI       0.170302 -0.034606 -0.154492  0.177429  0.245528  0.199757   
Aerosols -0.345247  0.014890  0.340238 -0.356155 -0.267809 -0.337055   

             CFC11     CFC12       TSI  Aerosols  
Year      0.569106  0.897012  0.170302 -0.345247  
Month    -0.013111  0.000675 -0.034606  0.014890  
MEI       0.069000  0.008286 -0.154492  0.340238  
CO2

In [48]:
#b
linear_regression_model_2 = smf.ols('Temp ~ MEI + TSI + Aerosols + N2O', data=train_data)
linear_regression_model_2_results = linear_regression_model_2.fit()

print(linear_regression_model_2_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.726
Model:                            OLS   Adj. R-squared:                  0.722
Method:                 Least Squares   F-statistic:                     184.9
Date:                Sun, 13 Oct 2024   Prob (F-statistic):           3.52e-77
Time:                        22:29:35   Log-Likelihood:                 266.64
No. Observations:                 284   AIC:                            -523.3
Df Residuals:                     279   BIC:                            -505.0
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -116.2269     20.223     -5.747      0.0

In [ ]:
#b.i
#N2O significance has improved compare to the last model.

#b.ii

Model 1 has a slightly higher R-Square (0.751) compared to Model 2 (0.726), indicating that it explains a bit more variance in temperature. However, Model 1 includes variables like CH4 and N2O, which are not statistically significant, reducing the model's reliability. Model 2, on the other hand, only includes statistically significant variables, making it more robust and easier to interpret. Despite the slightly lower R-Square Model 2 is more streamlined and reliable due to the significance of all its predictors. Overall, Model 1 offers better fit, but Model 2 provides a simpler, more interpretable solution with fewer, yet significant variables.

In [56]:
#c

test_data['PredictedTemp'] = (linear_regression_model_2_results.predict(X_test))

ss_residual = sum((test_data['Temp'] - test_data['PredictedTemp']) ** 2)
ss_total = sum((test_data['Temp'] - test_data['Temp'].mean()) ** 2)
r_squared_test = 1 - (ss_residual / ss_total)
print(r_squared_test)

-0.10237639062818116


#c

Negative R-squared suggests that the model's predictions on the test data are worse than if we had simply predicted the mean temperature for all instances. This indicates poor generalization performance and that the model is likely overfitting to the training data or the selected features (MEI, TSI, Aerosols, N2O) are not sufficient to explain the variability in the temperature in the test set.